src/main/java/sparksoniq/jsoniq/runtime/iterator/CommaExpressionIterator.java

```diff
@@ -17,59 +17,57 @@
  * Author: Stefan Irimescu
  *
  */
- package sparksoniq.jsoniq.runtime.iterator;
+package sparksoniq.jsoniq.runtime.iterator;
 
 import sparksoniq.jsoniq.item.Item;
 import sparksoniq.jsoniq.runtime.metadata.IteratorMetadata;
 import sparksoniq.semantics.DynamicContext;
 import sparksoniq.exceptions.IteratorFlowException;
 
+import java.util.ArrayList;
 import java.util.List;
 
 public class CommaExpressionIterator extends LocalRuntimeIterator {
 
+    private int _currentIndex;
+    private List<Item> results;
+
     public CommaExpressionIterator(List<RuntimeIterator> childIterators, IteratorMetadata iteratorMetadata) {
         super(childIterators, iteratorMetadata);
     }
 
     @Override
     public Item next() {
-        if(currentChild == null) {
-            currentChild = this._children.get(0);
-            this._children.forEach(c -> c.reset(this._currentDynamicContext));
-            currentChild.open(_currentDynamicContext);
-        }
-        if(currentChild.hasNext())
-            return currentChild.next();
-        else
-        {
-            currentChild.close();
-            if(this._children.indexOf(currentChild) == this._children.size() - 1)
-                throw new IteratorFlowException("Invalid next() call in Comma expression", getMetadata());
-            currentChild = this._children.get(_children.indexOf(currentChild) + 1);
-            currentChild.open(_currentDynamicContext);
-            return currentChild.next();
+        if (this._hasNext) {
+            return getResult();
         }
+        throw new IteratorFlowException("Invalid next() call in Comma expression", getMetadata());
     }
 
     @Override
-    public void open(DynamicContext context){
+    public void open(DynamicContext context) {
         super.open(context);
-        currentChild = null;
-    }
+        this._currentIndex = 0;
+        this.results = new ArrayList<>();
 
-    @Override
-    public void reset(DynamicContext context){
-        super.reset(context);
-        currentChild = null;
+        // this._children.forEach(c -> c.reset(this._currentDynamicContext));
```

Comment:
I understand the problem you are trying to solve but I think this is taking it a bit too far. Comma expressions, arrays and any Spark Iterator might be dealing with a very large amount of results and pre-computing all of the results in the open() method will potentially remove all the laziness resulting in significantly lower performance. @ghislainfourny what do you think?

Reply:
Ideally we would only peek at the first item in the open() method in order to return false for hasNext, instead of computing all the results, at least in some cases.

Reply:
Stefan has a good point -- we discussed this yesterday with Can.

The idea is to be as lazy as possible and *only* do the computations in open that are absolutely necessary (such as getting a lookahead of one item for hasNext() to return the correct result).

In some (rare) cases, it is necessary to compute everything in open. This is known in literature, and called "materialization". For example, this is the case with the reverse function: you can only return the first item by consuming the entire input sequence. This should be done only if absolutely necessary because of the semantics of the iterator.

This will partially be the case with local group by/order by execution, as the first tuple can only be returned after consuming the entire input stream.

Reply:
Regarding iterators with an underlying RDD: the design pretty much depends on the Spark API. If the Spark API allows us to "peek" at the RDD to see if it is empty, then it is not necessary to precompute in open() before haxNext() is called. If the Spark API allows getting the items one by one, this is also not necessary. If all Spark allows is "taking" all items at once, then materialization is necessary (but of course, in practice, this will only be done for small sequences, and an error can be thrown for bigger sequences).